### DISCUSSION TITLE
How do you only change the content of a custom node?

### DISCUSSION THREAD

**Discussion by Nefcanto** [xyflow/xyflow#4905] [Q&A]
I tried to create a custom node, and I got this error:

> [React Flow]: Couldn't create edge for source handle id: "null", edge id: e3-4. Help: https://reactflow.dev/error#008

Following that link I got to this example:

https://reactflow.dev/examples/nodes/custom-node

But that example uses Handles. I don't want to change anything. All I want is to show a `<div>{data.label}</div>`. That's all. How can I get that?

This is my `SimpleNode.jsx` content:

```
const Node = ({ data }) => {
    return <div className="border border-white ">
        <div>{data?.label}</div>
    </div>
})

export default Node
```

**Comment by bcakmakoglu** [comment]
> But that example uses Handles

If a node has edges going to it it *has to have* `<Handle>` components.

> All I want is to show a `<div>{data.label}</div>`.

You can very well do that but you're not gonna have any working edges for that node if all it does is display a label. It would be a node with no connections going to or from it.

**Reply by Nefcanto** [reply]
So, let me post my question this way. How can I only change the inner content of the default node?

**Reply by bcakmakoglu** [reply]
```ts
{
  id: "1",
  position: { x: -100, y: 100 },
  type: 'default', // can be omitted
  data: {
	// can be a string or a jsx element
 	label: <div>Node 1</div> 
  },
},
```

**Reply by Nefcanto** [reply]
But that does not work. Because when I style this div, I get a box inside another box (the default box).

**Reply by Nefcanto** [reply]
You know, the doc says:

> Creating your own nodes is as easy as creating a regular React component and passing them to nodeTypes.

It's not true. Creating our own nodes *is not* as easy as a regular React component. We need to handle everything related to handles and other stuff if we want to create it. In other words, to customize the inner of a node, we need to reimplement the entire node.

**Reply by bcakmakoglu** [reply]
> But that does not work. Because when I style this div, I get a box inside another box (the default box).

You asked how to change the inner content of a default node and I answered your question.
The default node(s) provide a "label" option which you can use to change the "inner content" of that node. 
Anything more customized than that requires you to create a custom node.

If you're just looking to change the border color of the wrapper of a node you can apply that as [`node.className`](https://reactflow.dev/api-reference/types/node)

```ts
{
  id: "1",
  position: { x: -100, y: 100 },
  className: 'border border-white',
  data: {
 	label: 'Node 1'
  },
},
```

> It's not true. Creating our own nodes is not as easy as a regular React component. We need to handle everything related to handles and other stuff if we want to create it. In other words, to customize the inner of a node, we need to reimplement the entire node.

Alas, I fail to see where the given statement that Custom Nodes are merely React components is false. 
They *are* regular React components.
They can be as simple as what you posted, with limitations that come with the simplicity, or as complex as you need them to be, like having n-Handles depending on some variable you use, like rows/columns for a table.

As with last time you seem to be under the impression that "arguing" will lead you somewhere but it won't.
All you do is antagonize people that spend their own time to try and help you and you come off as a very dis-likable person.

This will be the last time I reply to anything of yours, though I do wish you the very best and hope you succeed in your endeavors in the future.

**Reply by Nefcanto** [reply]
Thank you for being frank. I don't intend to argue. I just read the docs. Follow and encounter countless errors. The learning curve is too high. All I want to do is to customize the "inner content" of a node. But leave every behavior a node has. A node might jump to a grid, might pulsate when an edge comes near to it, might trigger some events for user's interaction. I'm not arguing with you. I ask for a simple requirement. It's not fair to re-implement the entire node and every behavior and outer-structuer it has (to work with edges) just because one decides to show some custom content. A good way could be to do something like:

```
const CustomNode = ({ data, ...rest }) => {
    return <div {...rest}>
        <div>Now this part is truly customizable</div>
    </div>
}
```

And that `...rest` would handle every default behavior and outer-structure a node has.

Please don't get angry. Please hear the pain a member of the community has, in order to integrate your library into a value deliverable to the end-user. That's all.

Again, thank you so much for your time. And please accept my apology if my tone seems harsh. I apologize.

**Reply by peterkogo** [reply]
Just look at the [default node code](https://github.com/xyflow/xyflow/blob/main/packages/react/src/components/Nodes/DefaultNode.tsx). Literally all you need to do is slap 2 Handle components inside your custom node. Like @bcakmakoglu already told you. 
There is no secret code. You don't need to reimplement anything.
